### DISCUSSION TITLE
Passkey Authentication with "Autofill UI"

### DISCUSSION THREAD

**Discussion by mweiershaeuser** [zitadel/zitadel#8867] [Q&A] [answered]
Hi everyone,

First of all a big thanks to the Zitadel team, my experience working with Zitadel so far has been pretty nice! :)

I have successfully implemented Passkey authentication with a custom Login UI for a university project, using a self-hosted Zitadel instance and now intended to augment the application with "Autofill UI" as suggested by W3C and the FIDO Alliance on passkeys.dev, which makes use of WebAuthn’s conditional mediation (see resources below). This feature would suggest usage of an existing Passkey to the user via autocomplete on the username input field and start the authentication from there, eliminating the need for any manual user input (including entering the username/email).

For this to work, the "Autofill UI" has to hand over the PublicKeyCredentialRequestOptions to navigator.credentials.get, before the username is known to the application. However, I have not been able to find a solution for this, since the Zitadel Session Service does not seem to be able to create a WebAuthn challenge without a preceding username check.

Therefore I wanted to ask, if I’m overlooking something or if this is not possible?

Thanks for any help and best regards!

Resources:
- [https://passkeys.dev/docs/reference/terms/#autofill-ui](https://passkeys.dev/docs/reference/terms/#autofill-ui)
- [https://passkeys.dev/docs/use-cases/bootstrapping/](https://passkeys.dev/docs/use-cases/bootstrapping/)

**Comment by hifabienne** [comment]
Hi @mweiershaeuser 
I think at the moment the auth flows in ZITADEL are implemented as identifier first, so you have to provide a username in the beginning.
But I am not 100% sure.
@livio-a @muhlemmer are you able to help here?

**Reply by muhlemmer** [reply]
I will read the docs and discuss it with @livio-a  on wednesday and come back with an answer.

**Reply by mweiershaeuser** [reply]
Thanks! :)

**Comment by hifabienne** [comment]
Hi @mweiershaeuser 
Sorry to sidetrack on this conversation. 😃
As I read about that you have implemented your custom Login UI, I would love to have a quick chat with you about your implementation and some feedback. If you are interested in that, please reach out to me. You can send me an email to fabienne@zitadel.com

**Comment by muhlemmer** [comment, marked as answer]
Hi @mweiershaeuser we are definitely interested in this idea. We would need to change the session API to allow for a random challenge to be set. On the management APIs of keys we will need to tie the UserID from a [resident key](https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/Resident_Keys.html). This way we can then set both user check and Webauthn as factors with a single interaction.

Let's create a feature request for this and then we can plan it.

**Reply by muhlemmer** [reply]
https://github.com/zitadel/zitadel/issues/8899

**Reply by mweiershaeuser** [reply]
Awesome, thanks for the feedback and for considering this!
